### ISSUE TITLE
Selecting maps with union_all and non-interpolated literals produces garbage results

### ISSUE THREAD

**Issue by liamwhite** [elixir-ecto/ecto#3191]
Possibly related to #3190

### Environment

* Elixir version (elixir -v): Elixir 1.9.4 (compiled with Erlang/OTP 22)
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): PostgreSQL 12
* Ecto version (mix deps): ecto 3.2.5 (Hex package) (mix)
* Database adapter and version (mix deps): postgrex 0.15.1 (Hex package) (mix)
* Operating system: Linux amd64

### Current behavior
[Code in question](https://github.com/derpibooru/philomena/blob/a3b3a38fef199fcd491494b7b4c5595f09200317/lib/philomena/interactions.ex#L14-L56):
```elixir
  def user_interactions(images, user) do
    ids =
      images 
      |> Enum.flat_map(fn
        nil -> []
        %{id: id} -> [id]
        enum -> Enum.map(enum, & &1.id)
      end)
      |> Enum.uniq()

    hide_interactions =
      ImageHide
      |> select([h], %{image_id: h.image_id, user_id: h.user_id, interaction_type: "hidden", value: ""})
      |> where([h], h.image_id in ^ids)
      |> where(user_id: ^user.id)

    fave_interactions =
      ImageFave
      |> select([f], %{image_id: f.image_id, user_id: f.user_id, interaction_type: "faved", value: ""})
      |> where([f], f.image_id in ^ids)
      |> where(user_id: ^user.id)

    upvote_interactions =
      ImageVote
      |> select([v], %{image_id: v.image_id, user_id: v.user_id, interaction_type: "voted", value: "up"})
      |> where([v], v.image_id in ^ids)
      |> where(user_id: ^user.id, up: true)

    downvote_interactions =
      ImageVote
      |> select([v], %{image_id: v.image_id, user_id: v.user_id, interaction_type: "voted", value: "down"})
      |> where([v], v.image_id in ^ids)
      |> where(user_id: ^user.id, up: false)

    [
      hide_interactions,
      fave_interactions,
      upvote_interactions,
      downvote_interactions
    ]
    |> union_all_queries()
    |> Repo.all()
  end

  defp union_all_queries([query]),
    do: query
  defp union_all_queries([query | rest]),
    do: query |> union_all(^union_all_queries(rest))
```
If the literal binaries above are not given as interpolated values, Ecto's output is not meaningful.

Contents of tables:
```
  ?column?  | image_id | user_id |  up   
------------+----------+---------+-------
 image_vote |        7 |       1 | true
 image_vote |        5 |       1 | true
 image_vote |        8 |       1 | false
 image_vote |       14 |       4 | true
 image_fave |        7 |       1 | 
 image_fave |        5 |       1 | 
 image_fave |       14 |       4 | 
```

Generated Ecto query:
```elixir
Ecto.Query<from i0 in Philomena.ImageHides.ImageHide,
 where: i0.image_id in ^[14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
 where: i0.user_id == ^1,
 union_all: (from i0 in Philomena.ImageFaves.ImageFave,
  where: i0.image_id in ^[14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
  where: i0.user_id == ^1,
  union_all: (from i0 in Philomena.ImageVotes.ImageVote,
  where: i0.image_id in ^[14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
  where: i0.user_id == ^1 and i0.up == true,
  union_all: (from i0 in Philomena.ImageVotes.ImageVote,
  where: i0.image_id in ^[14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
  where: i0.user_id == ^1 and i0.up == false,
  select: %{image_id: i0.image_id, user_id: i0.user_id, interaction_type: "voted", value: "down"}),
  select: %{image_id: i0.image_id, user_id: i0.user_id, interaction_type: "voted", value: "up"}),
  select: %{image_id: i0.image_id, user_id: i0.user_id, interaction_type: "faved", value: ""}),
 select: %{image_id: i0.image_id, user_id: i0.user_id, interaction_type: "hidden", value: ""}>
```

Generated SQL:
```elixir
{"SELECT i0.\"image_id\", i0.\"user_id\", 'hidden', '' FROM \"image_hides\" AS i0 WHERE (i0.\"image_id\" = ANY($1)) AND (i0.\"user_id\" = $2) UNION ALL (SELECT i0.\"image_id\", i0.\"user_id\", 'faved', '' FROM \"image_faves\" AS i0 WHERE (i0.\"image_id\" = ANY($3)) AND (i0.\"user_id\" = $4) UNION ALL (SELECT i0.\"image_id\", i0.\"user_id\", 'voted', 'up' FROM \"image_votes\" AS i0 WHERE (i0.\"image_id\" = ANY($5)) AND ((i0.\"user_id\" = $6) AND (i0.\"up\" = TRUE)) UNION ALL (SELECT i0.\"image_id\", i0.\"user_id\", 'voted', 'down' FROM \"image_votes\" AS i0 WHERE (i0.\"image_id\" = ANY($7)) AND ((i0.\"user_id\" = $8) AND (i0.\"up\" = FALSE)))))",
 [
   [14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
   1,
   [14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
   1,
   [14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
   1,
   [14, 13, 12, 11, 10, 8, 5, 4, 6, 3, 2, 1],
   1
 ]}
```
Returned results (nonsense):
```elixir
[
  %{image_id: 5, interaction_type: "hidden", user_id: 1, value: ""},
  %{image_id: 5, interaction_type: "hidden", user_id: 1, value: ""},
  %{image_id: 8, interaction_type: "hidden", user_id: 1, value: ""}
]
```
### Expected behavior
Result maps are expected to be given as
```elixir
[
  %{image_id: 5, interaction_type: "faved", user_id: 1, value: ""},
  %{image_id: 5, interaction_type: "voted", user_id: 1, value: "up"},
  %{image_id: 8, interaction_type: "voted", user_id: 1, value: "down"}
]
```
(and are when the binary value is marked interpolated using, for example, `^"faved"`)

**Comment by josevalim** [comment]
This has been fixed in Ecto v3.3. So please give it a try and let us know if it works as expected. Thanks!

**Comment by liamwhite** [comment]
Seems to be, thanks.
